source/common/config/grpc_mux_impl.cc

```diff
@@ -14,6 +14,33 @@
 namespace Envoy {
 namespace Config {
 
+namespace {
+class AllMuxesState {
+public:
+  void insert(GrpcMuxImpl* mux) {
+    absl::WriterMutexLock locker(&lock_);
+    muxes_.insert(mux);
+  }
+
+  void erase(GrpcMuxImpl* mux) {
+    absl::WriterMutexLock locker(&lock_);
+    muxes_.erase(mux);
+  }
+
+  void shutdownAll() {
+    absl::WriterMutexLock locker(&lock_);
+    for (auto& mux : muxes_) {
+      mux->shutdown();
+    }
+  }
+
+private:
+  absl::flat_hash_set<GrpcMuxImpl*> muxes_ ABSL_GUARDED_BY(lock_);
+  absl::Mutex lock_;
+};
+using AllMuxes = ThreadSafeSingleton<AllMuxesState>;
```

Comment:
Technically, does this even need locking? Isn't this all accessed only from the main thread?

Reply:
I wasn't sure, and I especially wasn't sure in the context of what tests do, so I just added the locking to make sure it won't make things worse than they are now.

Reply:
OK, I'm pretty sure it's not needed, so maybe also add a comment/TODO about that?